enhancements/storage/storage-hypershift.md

```diff
@@ -0,0 +1,1050 @@
+   guest cluster.
+
+All storage operators (cluster-storage-operator, csi-driver-operator, and csi-snapshot-controller-operator) behave the
+same as in standalone OCP, they just create / watch / sync objects in two clusters instead of one. This allows us to
+reuse the existing operators.
+
+<!--
+Explain how the user will use the feature. Be detailed and explicit.
+Describe all of the actors, their roles, and the APIs or interfaces
+involved. Define a starting state and then list the steps that the
+user would need to go through to trigger the feature described in the
+enhancement. Optionally add a
+[mermaid](https://github.com/mermaid-js/mermaid#readme) sequence
+diagram.
+
+Use sub-sections to explain variations, such as for error handling,
+failure recovery, or alternative outcomes.
+
+For example:
+
+**cluster creator** is a human user responsible for deploying a
+cluster.
+
+**application administrator** is a human user responsible for
+deploying an application in a cluster.
+
+1. The cluster creator sits down at their keyboard...
+2. ...
+3. The cluster creator sees that their cluster is ready to receive
+   applications, and gives the application administrator their
+   credentials.
+
+-->
+
+#### Variation [optional]
+
+N/A ?
+
+<!--
+If the cluster creator uses a standing desk, in step 1 above they can
+stand instead of sitting down.
+
+See
+https://github.com/openshift/enhancements/blob/master/enhancements/workload-partitioning/management-workload-partitioning.md#high-level-end-to-end-workflow
+and https://github.com/openshift/enhancements/blob/master/enhancements/agent-installer/automated-workflow-for-agent-based-installer.md for more detailed examples.
+
+-->
+
+### API Extensions
+
+No API changes.
+
+<!--
+API Extensions are CRDs, admission and conversion webhooks, aggregated API servers,
+and finalizers, i.e. those mechanisms that change the OCP API surface and behaviour.
+
+- Name the API extensions this enhancement adds or modifies.
+- Does this enhancement modify the behaviour of existing resources, especially those owned
+  by other parties than the authoring team (including upstream resources), and, if yes, how?
+  Please add those other parties as reviewers to the enhancement.
+
+  Examples:
+  - Adds a finalizer to namespaces. Namespace cannot be deleted without our controller running.
+  - Restricts the label format for objects to X.
+  - Defaults field Y on object kind Z.
+
+Fill in the operational impact of these API Extensions in the "Operational Aspects
+of API Extensions" section.
+-->
+
+### Implementation Details/Notes/Constraints [optional]
+
+<!-- What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+-->
+
+#### csi-snapshot-controller
```

Comment:
If you combine the diagrams in this section it would be clearer that the cluster-csi-snapshot-controller-operator is reconciling resources in both clusters (deployments in the management-cluster and a bunch of other things in the guest-cluster) at the same time.

Could we run 2 copies of that operator separately with command line flags to enable/disable certain reconciliation steps so that each binary only had to reconcile resources in 1 cluster (and therefore only deal with one kube client)?

Reply:
How would these two operators agree on ClusterOperator.Status? IMO it would be more complicated / confusing than running 1 operator with 2 kubeconfigs.

Reply:
Why would they need to agree on status? Why not report status separately?

Reply:
Do you mean two OperatorStatus CRs for csi-snapshot-controller? Yes, we could have them, but I don't see the benefit. Yes, one operator instance will have just one kubeconfig, but the source code will still have code paths with different kubeconfigs.